enhancements/namespace-auto-pruning.md

```diff
@@ -0,0 +1,170 @@
+
+1. Should there be a dry run feature to test the policy?
+1. Should this only be implemented in the new UI and not the old UI?
+1. Can a user configure multiple policies per namespace?
+1. Can a user specify tags or repositories by pattern?
+
+## Summary
+
+In order for a container registry to run cost effectively unused resources should be deleted to reclaim storage space. This task should be done automatically in Quay by deleting unused tags through a given policy. The policy used should be configured by the user and a variety policy options should be given. Examples of initial policies would be prune tags older than X number of days or prune tags when more than 100 are present.
+
+## Motivation
+
+Reclaiming space is essential in running a cost efficient registry. Having this done automatically within Quay removes the need for manual intervention or custom software.
+
+### Goals
+
+* User is able to provide a time based policy for auto-pruning tags within a namespace
+* User is able to provide a tag number based policy for auto-pruning tags within a namespace
+* User is able to specify whether tags go into the time machine or are immediately deleted
+* The correct tags are deleted when the policy is applied
+* Pruned tags are tracked via usage logs
+
+### Non-Goals
+
+* Allowing users to specify multiple policies per namespace
+* Applying policies based on repo or tag name pattern matching
+* Allowing policy configurations at the repository level
+
+## Proposal
+
+The details in this proposal is only for namespace auto-pruning policies but designed in a way to support repository auto-pruning policies in the future. This proposal is structured in three parts: how policies are managed, how the policies are structured in the DB, and how the policies will be executed via the auto-prune worker.
+
+### Managing Policies
+
+A user is able to configure auto-pruning within the namespace settings. To start, there will be two auto-pruning options:
+* Prune by number of tags, provided the number of tags
+* Prune tags older than date, provided the number of days 
+
+For this implementation users will only be able to configure one policy per namespace. To interface with this new UI there will be a new set of CRUD endpoints for managing auto-pruning policies. The endpoints returning lists will be paginated.
+
+**Create policy**
+```
+POST /api/v1/organization/{organization}/autoprunepolicy
+By date:
+{
+    pruneBy: "creation_date",
+    olderThan: "2w"  
+}
+By number of tags:
+{
+    pruneBy: "number_of_tags",
+    numTags: 100  
+}
+```
+
+**Update policy**
+```
+PATCH /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+{
+    fieldToBeUpdated: value,
+}
+```
+
+**List policies**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy
+```
+
+**Get policy**
+```
+GET /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
+
+**Delete policy**
+```
+DELETE /api/v1/organization/{organization}/autoprunepolicy/{policyid}
+```
```

Comment:
What is the unit of work of these policies, tags or manifests?
For example, if multiple tags of different ages point to a single manifest, then these policy wouldn't do much in terms of reclaiming storage space.

Reply:
Tags. You're correct but I think the tags pointing to the same manifest would be created at similar times meaning they would be pruned by a similar time as well. This would probably be something to document for the user.